src/main/java/net/minecraftforge/client/ClientCommandHandler.java

```diff
@@ -0,0 +1,312 @@
+import com.mojang.brigadier.tree.RootCommandNode;
+import net.minecraft.ChatFormatting;
+import net.minecraft.Util;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.player.LocalPlayer;
+import net.minecraft.commands.CommandRuntimeException;
+import net.minecraft.commands.CommandSourceStack;
+import net.minecraft.commands.SharedSuggestionProvider;
+import net.minecraft.commands.synchronization.SuggestionProviders;
+import net.minecraft.network.chat.ClickEvent;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.ComponentUtils;
+import net.minecraft.network.chat.HoverEvent;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.TextComponent;
+import net.minecraft.network.chat.TranslatableComponent;
+import net.minecraftforge.client.event.RegisterClientCommandsEvent;
+import net.minecraftforge.common.MinecraftForge;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+public class ClientCommandHandler
+{
+    private static final Logger LOGGER = LogManager.getLogger();
+    private static CommandDispatcher<CommandSourceStack> commands = null;
+
+    /*
+     * For internal use
+     * 
+     * Merges command dispatcher use for suggestions to the command dispatcher used for client commands so they can be sent to the server, and vice versa so client commands appear
+     * with server commands in suggestions
+     */
+    public static CommandDispatcher<SharedSuggestionProvider> mergeServerCommands(CommandDispatcher<SharedSuggestionProvider> serverCommands)
+    {
+        CommandDispatcher<CommandSourceStack> commandsTemp = new CommandDispatcher<>();
+        MinecraftForge.EVENT_BUS.post(new RegisterClientCommandsEvent(commandsTemp));
+
+        // Copies the client commands into another RootCommandNode so that redirects can't be used with server commands
+        commands = new CommandDispatcher<>();
+        copy(commandsTemp.getRoot(), commands.getRoot());
+
+        // Copies the server commands into another RootCommandNode so that redirects can't be used with client commands
+        RootCommandNode<SharedSuggestionProvider> serverCommandsRoot = serverCommands.getRoot();
+        CommandDispatcher<SharedSuggestionProvider> newServerCommands = new CommandDispatcher<>();
+        copy(serverCommandsRoot, newServerCommands.getRoot());
+
+        // Copies the server side commands into a temporary server side commands root node to be used later without the client commands
+        RootCommandNode<SharedSuggestionProvider> serverCommandsCopy = new RootCommandNode<>();
+        mergeCommandNode(newServerCommands.getRoot(), serverCommandsCopy, new IdentityHashMap<>(),
+                Minecraft.getInstance().getConnection().getSuggestionsProvider(), (suggestions) -> 0, (suggestions) -> null);
```

Comment:
Lambda parameters should be better named:
```suggestion
                Minecraft.getInstance().getConnection().getSuggestionsProvider(), (context) -> 0, (suggestions) -> null);
```